Add import status endpoint and progress options

joshuatf opened the pull request:
Fixes #1850

Adds an endpoint to query the current import progress and oldest date of import.  The date will be set to `-1` if `days` are not specific initially.

Also adds an endpoint to query potential totals prior to import.

### Thoughts/notes
* Opted to use options here instead of attempting to query pending/completed batches.  The latter seems like it would get messy quickly and is prone to inaccurate results with completed results.
* Orders total is not calculated until after the customers batch is finished since it's a dependent action.  If we want to show orders total while customers are being processed, we will probably need to create both customer and order batches initially so that we can calculate totals and set the orders batch to pending until the customers batch is complete.
* Imports should still be limited to one import at a time.  Will outline a separate issue to tackle this.

### Screenshots
<img width="365" alt="Screen Shot 2019-05-08 at 3 56 39 PM" src="https://user-images.githubusercontent.com/10561050/57359381-2aec0280-71aa-11e9-84d0-30c8af330d64.png">

### Detailed test instructions:

1.  Start an import by posting to the `/wp-json/wc/v4/reports/import` endpoint.
2.  Send a get request to `/wp-json/wc/v4/reports/import/status`.
3.  Check that totals and progress count are reflected in the endpoint.
4.  Send a get request to `/wp-json/wc/v4/reports/import/totals` with `day` and `skip_existing` params.
5.  Make sure that total numbers returned are accurate.

joshuatf committed:
Update import progress during batches

joshuatf committed:
Add import status route for querying import progress

Aljullu commented on includes/api/class-wc-admin-rest-reports-import-controller.php:
```diff
@@ -249,4 +261,27 @@ public function delete_imported_items( $request ) {
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_import_status( $request ) {
+		$delete = WC_Admin_Reports_Sync::delete_report_data();
```
I don't understand this line. Is it a leftover from another function?

Aljullu reviewed (commented):
Thanks for working on this @joshuatf! Just a couple of comments/questions:

> Orders total is not calculated until after the customers batch is finished since it's a dependent action. If we want to show orders total while customers are being processed, we will probably need to create both customer and order batches initially so that we can calculate totals and set the orders batch to pending until the customers batch is complete.

According to the designs it looks like we should display the totals even before starting the import:

![image](https://user-images.githubusercontent.com/3616980/57367971-56f78b80-718a-11e9-80e9-94eb472d7e37.png)

Not sure if that's possible or we should just hide the progress bars when the import hasn't started yet.

joshuatf committed:
Update import totals before import start

joshuatf committed:
Move wc admin notes to separate action scheduler group

joshuatf committed:
Add import totals route to query customer and order totals

joshuatf committed:
Add wc helper queue job processing method

joshuatf committed:
Add import status and total tests

joshuatf commented:
Made some changes to this PR:

* Added `is_importing` to the status route.
* Added a separate totals endpoint (testing instructions above) to query totals prior to importing.
* Set totals for current import prior to beginning import.
* Added tests for totals and status endpoints.

@Aljullu this is ready for another look if you have time.

Aljullu commented on includes/notes/class-wc-admin-notes.php:
```diff
@@ -131,7 +136,21 @@ public static function schedule_unsnooze_notes() {
+
+		if ( is_a( $store, 'WC_Admin_ActionScheduler_WPPostStore' ) ) {
+			// If we're using our data store, call our bespoke deletion method.
+			$store->clear_pending_wcadmin_actions();
```
This method [seems to clear import actions](https://github.com/woocommerce/woocommerce-admin/blob/master/includes/class-wc-admin-actionscheduler-wppoststore.php#L42), right? Shouldn't we create a different method which only cleared note actions?

joshuatf replied:
Ya, I think you're right and we should actually extract the actions out of that method to make this reusable.  Updated in https://github.com/woocommerce/woocommerce-admin/pull/2189/commits/13753b7e47869c3b55ead67b34dad87757d99945

Aljullu reviewed (commented):
Thanks for the last changes @joshuatf!

This is testing great and code looks good to me, I just added a small question in the code.

joshuatf committed:
Separate action type clearing for imports and notes

joshuatf commented:
Thanks again for the review @Aljullu!  Added an update to separate out action clearing for notes and imports.

Aljullu reviewed (approved):
Thanks for the changes @joshuatf! LGTM

rrennick commented on includes/class-wc-admin-actionscheduler-wppoststore.php:
```diff
@@ -36,23 +36,15 @@ protected function create_post_array( ActionScheduler_Action $action, DateTime $
-			WC_Admin_Reports_Sync::SINGLE_ORDER_IMPORT_ACTION,
-		);
-
 		foreach ( $action_types as $action_type ) {
```
Since this is a public function `$action_types` should probably be cast to an array.

joshuatf committed:
Typecast action types as array in clear_pending_wcadmin_actions()
